neps/nep-0399.md

```diff
@@ -0,0 +1,340 @@
+   We can't conditionally enable FlatStorage for some blocks and disable it for other, because 
+   the fees we charge must be consistent. 
+   
+The mapping of key value pairs FlatStorage stored on disk matches the state at some block. 
+We call this block the head of flat storage, or the flat head. There are two natural options 
+for which block should be the flat head, the chain head, or the last final block. Although
+both options could work, the implementation we propose will use the last final block because 
+it is simpler.
+
+To support key value lookups for other blocks that are not the flat head, FlatStorage will 
+also store key value changes(deltas) per block for these blocks.  
+We call these deltas FlatStorageDelta (FSD). Let’s say the flat storage head is at block h, 
+and we are applying transactions based on block h’. Since h is the last final block, 
+h is an ancestor of h'. To access the state at block h', we need FSDs of all blocks between h and h'.
+Note that all these FSDs must be stored in memory, otherwise, the access of FSDs will trigger 
+more disk reads and we will have to set storage key read fee higher. 
+
+However, the consensus algorithm doesn’t provide any guarantees in the distance of blocks 
+that we need to process since it could be arbitrarily long for a block to be finalized. 
+To solve this problem, we make another proposal (TODO: attach link for the proposal) to 
+set gas limit to zero for blocks with height larger than the latest final block’s height + X. 
+If the gas limit is set to zero for a block, it won't contain any transactions or receipts, 
+and FlatStorage won't need to store the delta for this block.
+With this change, FlatStorage only needs to store FSDs for blocks with height less than the latest 
+final block’s height + X. And since there can be at most one valid block per height, 
+FlatStorage only needs to store at most X FSDs in memory. 
+
+### FSD size estimation
+To set the value of X, we need to see how many block deltas can fit in memory.
+
+We can estimate FSD size using protocol fees. 
+Assume that flat state stores a mapping from keys to value refs. 
+Maximal key length is ~2 KiB which is the limit of contract data key size. 
+During wasm execution, we pay `wasm_storage_write_base` = 64 Ggas per call and 
+`wasm_storage_write_key_byte` = 70 Mgas per key byte. 
+In the extreme case it means that we pay `(64_000 / 2 KiB + 70) Mgas ~= 102 Mgas` per byte. 
+Then the total size of keys changed in a block is at most 
+`block_gas_limit / gas_per_byte * num_shards = (1300 Tgas / 102 Mgas) * 4 ~= 50 MiB`.
+
+To estimate the sizes of value refs, there will be at most 
+`block_gas_limit / wasm_storage_write_base * num_shards
+= 1300 Tgas / 64 Ggas * 4 = 80K` changed entries in a block. 
+Since one value ref takes 40 bytes, limit of total size of changed value refs in a block 
+is then 3.2 MiB. 
+
+To sum it up, we will have < 54 MiB for one block, and ~1.1 GiB for 20 blocks.
```

Comment:
- How do we ensure that FSDs are kept in memory?  Will we use a memory only data struct for them like hash maps?
- Do we know how many memory accesses we need to check if a worst case FSD of 54 MiB has a given key or not?

Reply:
- Yes. They will be stored in `FlatStateDelta`. They are also persisted on disk, but we don't read from disk during block processing. We only read from disk for this column when node first starts and FlatStorage loads deltas for all blocks after the flat head into memory. 
- Most of the times twice because in most case there is no fork, and the chain head should be at last final block + 2. Worst case, it can be up to he number of blocks between the last final block to this block. If we implement the idea of setting gas limit to zero for blocks that are further than X blocks from the final block, then the number of in memory hashmap lookup will be X.